offchain/eth-input-reader/src/eth_input_reader.rs

```diff
@@ -114,24 +136,21 @@ pub async fn start(
 #[instrument(level = "trace", skip_all)]
 #[allow(clippy::too_many_arguments)]
 async fn process_block(
+    env: &Arc<StateFoldEnvironment<InputProvider, Mutex<UserData>>>,
     block: &Block,
-
-    state_server: &impl StateServer<
-        InitialState = RollupsInitialState,
-        State = RollupsState,
-    >,
     initial_state: &RollupsInitialState,
-
     context: &mut Context,
     machine_driver: &mut MachineDriver,
-
     broker: &(impl BrokerSend + BrokerReceive),
 ) -> Result<(), EthInputReaderError> {
     trace!("Querying rollup state");
-    let state = state_server
-        .query_state(initial_state, block.hash)
-        .await
-        .context(StateServerSnafu)?;
+    let state = RollupsState::get_state_for_block(
+        &Arc::new(initial_state.to_owned()),
+        block,
+        env,
+    )
+    .await
+    .expect("should get state");
```

Comment:
At three arguments to just call this function (`initial_state`, `env`, `block`), maybe you should get the state before calling this function, and just pass the state.

I'd even rethink the role of this whole function (`process_block`). The main loop used to be on a different function too, separating it from setup. Now that they're joined, maybe there's a better way to do things.

I personally like smaller functions which do one think for readability and testability. YMMV. Whatever you prefer is fine, just evaluate if this balance is to your liking.